### ISSUE TITLE
Entities of generated metamodels don't understand accept:

### ISSUE THREAD

**Issue by enwiro** [moosetechnology/Famix#303]
If we want to apply a visitor on the entities of a generated meta models, it is not possible to do so since it does not implement the accept: method.
![image](https://user-images.githubusercontent.com/58632394/118998553-d3ccbb00-b989-11eb-8d30-e4b121ef3b27.png)

**Comment by NicolasAnquetil** [comment]
Yes, it would be useful, not only for visitors, but to implement double-dispatch mechanism.

It would be nice that all composing traits implements #accept: with a specific #visitXYZ: call.
That way, we would be able to implement behavior specific to all traits in the visitors/double-dispatch.

However, each concrete class typically uses several traits and obviously, it cannot have several #accept: methods.
So it is not entirely clear how to do this properly ...

**Comment by NicolasAnquetil** [comment]
Thinking about it, may be the visitor pattern is not well adapted here because the goal would be to have a visitor accepting any new meta-model which does not seem doable

I am proposing to close the issue as won't do

**Comment by enwiro** [comment]
Yes, I hadn't seen it from that angle. All these behaviors are not even necessarily generic, so it is complicated to imagine a common visitor (especially for new meta-models).

I agree to close this topic.

**Comment by NicolasAnquetil** [comment]
OK, didn't close it yet :-)
Still thinking about it has the double dispatch pattern/visitor is really useful and it seems to fit so well the data we have.

One possibility would be that every Famix trait has its own #accept: method.
Everything that uses Famix traits would be generated with an #accept: method that would call  #visitXYZ: for all the Famix traits used by this thing, even traits used transitively.

That way we don't need to know all the entities implemented (FamixJavaMethod, FamixPharoMethod, FamixCppMethod, FamixPythonMethod, ...), but "only" all the traits which are much less (but still a lot).

Visitors would be more generic also, as they would #visitTMethod: instead of #visitJavaMethod: , ...
Visitors could also look at specific aspects of the models #visitTWithInheritance: , #visitTHasVisibility , ...

The result will be huge #accept: methods, but if that's what is needed ... 

This seems like a job for the generator of meta-models

**Comment by Ducasse** [comment]
We could also have a reflective one

```
accept: aFoo

   ^   aFoo perform: 'visit', XXXX asMutator with: self
```

now XXX is the key point.
Do we put the trait name?
Or something else

**Comment by NicolasAnquetil** [comment]
I needed it for FASTFortran, so I implemented a small tool that does it on an already existing metamodel, see: [https://github.com/moosetechnology/Famix/tree/visitingInfrastructure](https://github.com/moosetechnology/Famix/tree/visitingInfrastructure).

The idea is:
- start from root entity of the model
- for all subclasses generate an #accept: as an extension of a given 'XYZ-Visitor' package (that's the easy part, but the reflective solution of Steph would also work)
- in this package there must be an AbstractVisitor class (didn't bother check if it exists and create it otherwise)
- add to it a `#visitXyz: aNode` for each Xyz subclass of the root entity
  - visit method starts by doing a `self visitSuperOfXyz: aNode.` (call to the visit of the superclass, not trait)
  - then for each child instance variable having an accessor in the class do `aNode childAccessor accept: self`
  - child accessors are the methods with same name as an instance variable that don't have the `<container>` pragma

Not entirely satisfying as there is a risk of recursive loop (a type visits its #typedEntities which is a variable, and the variable visits its #declaredType which is the type).
But it did create me a base visiting infrastructure for a complex metamodel (Fortran: >100 entites and still incomplete)

**Comment by NicolasAnquetil** [comment]
Still needed.

For now I propose that the generator creates a basic `#accept:` method in each entity that calls `#visitTheClassOfTheEntity:  self`
That would be useful to anything that needs double-dispatch (not just visitors)

**Comment by NicolasAnquetil** [comment]
same as #938
